File: business_logic.ts
import _ from "lodash";
import type {CellValue, GameReport, Player} from "./domain";
import {Option, Exit, Cause} from "effect";

/**
 * Transposes a square matrix.
 * @param {CellValue[][]} board - A 2D array representing the tic-tac-toe board
 * @returns {CellValue[][]} The transposed matrix
 */
const transpose = (board: CellValue[][]): CellValue[][] => {
    return _.zip(...board) as CellValue[][];
};

/**
 * Collects the main diagonal of a square matrix.
 * @param {CellValue[][]} board - A 2D array representing the tic-tac-toe board
 * @returns {CellValue[]} An array containing the elements of the main diagonal
 */
const collectDiagonal = (board: CellValue[][]): CellValue[] => board.map((row, index) => row[index]) as CellValue[];

/** Collects the anti-diagonal of a square matrix.
 * @param {CellValue[][]} board - A 2D array representing the tic-tac-toe board
 * @returns {CellValue[]} An array containing the elements of the anti-diagonal
 */
const collectAntiDiagonal = (board: CellValue[][]): CellValue[] => {
    return board.map((row, index) => row[row.length - 1 - index]) as CellValue[];
};

/**
 * Validates the input is a non-empty square matrix.
 * @param {CellValue[][]} board - A 2D array representing the tic-tac-toe board
 * @returns {Effect.Exit} An `Effect.Exit` object indicating either success, or the failure cause
 */
const validateLegalBoard = (board: CellValue[][]): Exit.Exit<CellValue[][], string> => {
    if (
        !Array.isArray(board) ||
        board.length === 0 ||
        !board.every((sub) => Array.isArray(sub) && sub.length === board.length)
    ) {
        return Exit.failCause(Cause.fail("Input must be a non-empty square matrix"));
    } else return Exit.succeed(board);
};

/**
 * @param {CellValue[][]} board - A 2D array representing the tic-tac-toe board
 * @returns {Record<CellValue, number>} An object counting the occurrences of each player's tokens and empty cells
 */
const tokenCounter = (board: CellValue[][]): Record<CellValue, number> => {
    return _.countBy(_.flattenDeep(board), (cell) => (cell === "X" || cell === "O" ? cell : ".")) as Record<
        CellValue,
        number
    >;
};

/**
 * @param {CellValue[]} subArray - A 1D array representing a row, column, or diagonal of the tic-tac-toe board
 * @returns {Option.Option<Player>} An `Option.Option` object indicating the winner, if any
 */
const subArrayWinner = (subArray: CellValue[]): Option.Option<Player> => {
    if (subArray.every((elem) => elem === subArray[0]) && subArray[0] !== ".") {
        return Option.some(subArray[0] as Player);
    }
    return Option.none();
};

export const ticTacToeVerifier = (board: CellValue[][]): GameReport => {
    // 1. Validate the board is a non-empty square matrix
    if (Exit.isFailure(validateLegalBoard(board))) {
        return {
            validity: "Illegal",
            cause: Option.some("Input must be a non-empty square matrix"),
            state: Option.none(),
            winner: Option.none(),
        };
    }

    // 2. Count the tokens for each player and empty cells, and validate the counts are legal
    const tokenCounts = tokenCounter(board);
    const xCount = tokenCounts["X"] || 0;
    const oCount = tokenCounts["O"] || 0;

    if (xCount < oCount || xCount > oCount + 1) {
        return {
            validity: "Invalid",
            cause: Option.some(
                "The number of 'X' tokens must be equal to or exactly one more than the number of 'O' tokens."
            ),
            state: Option.none(),
            winner: Option.none(),
        };
    }

    // 3. Check for winners in rows, columns, and diagonals
    const linesToCheck = [
        ...board, // rows
        ...transpose(board), // columns
        collectDiagonal(board), // main diagonal
        collectAntiDiagonal(board), // anti-diagonal
    ];

    const winners = linesToCheck
    .map(subArrayWinner)
    .filter(Option.isSome)
    .map((opt) => opt.value);

    const uniqueWinners = Array.from(new Set(winners));

    if (uniqueWinners.length > 1) {
        return {
            validity: "Invalid",
            cause: Option.some("Both players cannot win simultaneously."),
            state: Option.none(),
            winner: Option.none(),
        };
    }

    if (uniqueWinners.length === 1) {
        const winner = uniqueWinners[0];
        if (winner === "X" && xCount !== oCount + 1) {
            return {
                validity: "Invalid",
                cause: Option.some("Invalid number of 'X' tokens for 'X' to win."),
                state: Option.none(),
                winner: Option.none(),
            };
        }
        if (winner === "O" && xCount !== oCount) {
            return {
                validity: "Invalid",
                cause: Option.some("Invalid number of 'O' tokens for 'O' to win."),
                state: Option.none(),
                winner: Option.none(),
            };
        }
        return {
            validity: "Valid",
            cause: Option.none(),
            state: winner === "X" ? Option.some("XWon") : Option.some("OWon"),
            winner: Option.some(winner) as Option.Option<Player>,
        };
    }

    // 4. If no winners, determine if the game is a draw or still in progress
    const emptyCount = tokenCounts["."] ?? 0;
    if (emptyCount === 0) {
        return {
            validity: "Valid",
            cause: Option.none(),
            state: Option.some("Draw"),
            winner: Option.none(),
        };
    } else {
        return {
            validity: "Valid",
            cause: Option.none(),
            state: Option.some("InProgress"),
            winner: Option.none(),
        };
    }
};

export const prettyPrintGameReport = (report: GameReport): string => {
    const {validity, state, winner} = report;

    if (validity === "Illegal") {
        return (
            "The board is illegal: " +
            Option.match(report.cause, {
                onNone: () => "Unknown cause",
                onSome: (cause) => cause,
            })
        );
    } else if (validity === "Invalid") {
        return (
            "The board is invalid: " +
            Option.match(report.cause, {
                onNone: () => "Unknown cause",
                onSome: (cause) => cause,
            })
        );
    } else {
        const stateStr = Option.match(state, {
            onNone: () => "Unknown State",
            onSome: (s) => s,
        });
        const winnerStr = Option.match(winner, {
            onNone: () => "No Winner",
            onSome: (w) => w,
        });
        return `The board is valid. State: ${stateStr}. Winner: ${winnerStr}.`;
    }
};

File: business_logic.py
from typing import Literal

from domain import Validity


def validate_board(board) -> Literal[Validity.ILLEGAL] | None:
    """
    Validate the input is a legal Tic-Tac-Toe board, i.e., a square matrix

    Args:
        board (list[list[Cell]]): A 2D list representing the Tic-Tac-Toe board
    Returns:
        Validity.ILLEGAL | None: If the board is illegal, return `Validity.ILLEGAL`; otherwise, return `None`
    """
    from domain import Validity

    if not isinstance(board, list) or len(board) == 0:
        return Validity.ILLEGAL
    n = len(board)
    for row in board:
        if not isinstance(row, list) or len(row) != n:
            return Validity.ILLEGAL
        for cell in row:
            if cell not in ("X", "O", "-"):
                return Validity.ILLEGAL
    return None


def transpose(board):
    """
    Transpose the input board, i.e., swap rows and columns

    Args:
        board (list[list[Cell]]): A 2D list representing the Tic-Tac-Toe board
    Returns:
        list[list[Cell]]: The transposed board
    """
    return [list(row) for row in zip(*board)]


def get_diagonals(board):
    """
    Get the two diagonals of the input board

    Args:
        board (list[list[Cell]]): A 2D list representing the Tic-Tac-Toe board
    Returns:
        list[list[Cell]]: A list containing the two diagonals of the board
    """
    n = len(board)
    return [[board[i][i] for i in range(n)], [board[i][n - 1 - i] for i in range(n)]]


def tic_tac_toe_verification(board):
    """
    Verify the input board is a valid Tic-Tac-Toe board and determine the game result

    Args:
        board (list[list[Cell]]): A 2D list representing the Tic-Tac-Toe board
    Returns:
        VerificationResult: A named tuple containing the legality, validity, game result, and cause of invalidity (if any)
    """
    from domain import GameResult, Validity, VerificationResult

    # 1. Check legality
    illegal = validate_board(board)
    if illegal is not None:
        return VerificationResult(
            legality=Validity.ILLEGAL,
            cause="Board is not a square or contains illegal values.",
        )

    n = len(board)
    flat = [cell for row in board for cell in row]
    x_count = flat.count("X")
    o_count = flat.count("O")
    dash_count = flat.count("-")

    # 2. Find winners
    def winner_lines():
        lines = []
        lines.extend(board)
        lines.extend(transpose(board))
        lines.extend(get_diagonals(board))
        return lines

    win_lines = winner_lines()
    x_win = any(all(cell == "X" for cell in line) for line in win_lines)
    o_win = any(all(cell == "O" for cell in line) for line in win_lines)

    # 3. Check for multiple winners
    if x_win and o_win:
        return VerificationResult(
            legality=Validity.LEGAL,
            validity=Validity.INVALID,
            cause="Both players cannot win.",
        )

    # 4. Check move counts
    # X always goes first
    if not (x_count == o_count or x_count == o_count + 1):
        return VerificationResult(
            legality=Validity.LEGAL,
            validity=Validity.INVALID,
            cause="Invalid number of moves.",
        )

    # 5. Check win conditions
    if x_win:
        if x_count != o_count + 1:
            return VerificationResult(
                legality=Validity.LEGAL,
                validity=Validity.INVALID,
                result=GameResult.X_WINS,
                cause="X must have one more move than O if X wins.",
            )
        return VerificationResult(
            legality=Validity.VALID, validity=Validity.VALID, result=GameResult.X_WINS
        )
    if o_win:
        if x_count != o_count:
            return VerificationResult(
                legality=Validity.VALID,
                validity=Validity.INVALID,
                result=GameResult.O_WINS,
                cause="O must have same number of moves as X if O wins.",
            )
        if dash_count == 0:
            return VerificationResult(
                legality=Validity.VALID,
                validity=Validity.INVALID,
                result=GameResult.O_WINS,
                cause="There must be at least one empty cell if O wins.",
            )
        return VerificationResult(
            legality=Validity.VALID, validity=Validity.VALID, result=GameResult.O_WINS
        )

    # 6. Draw
    if dash_count == 0:
        if x_count != o_count + 1:
            return VerificationResult(
                legality=Validity.VALID,
                validity=Validity.INVALID,
                result=GameResult.DRAW,
                cause="Drawn game must have X with one more move than O.",
            )
        return VerificationResult(
            legality=Validity.VALID, validity=Validity.VALID, result=GameResult.DRAW
        )

    # 7. Ongoing
    return VerificationResult(
        legality=Validity.VALID, validity=Validity.VALID, result=GameResult.ONGOING
    )


def pretty_print_results(verification_result):
    """
    Pretty print the verification result

    Args:
        verification_result (VerificationResult): A named tuple containing the legality, validity, game result, and cause of invalidity (if any)
    """
    from domain import GameResult, Validity

    if verification_result.legality == Validity.ILLEGAL:
        print("Board is ILLEGAL.")
        if verification_result.cause:
            print(f"Cause: {verification_result.cause}")
        return
    if verification_result.validity == Validity.INVALID:
        print("Board is INVALID.")
        if verification_result.cause:
            print(f"Cause: {verification_result.cause}")
        return
    if verification_result.result == GameResult.DRAW:
        print("Game ended in a DRAW.")
    elif verification_result.result == GameResult.X_WINS:
        print("X wins!")
    elif verification_result.result == GameResult.O_WINS:
        print("O wins!")
    elif verification_result.result == GameResult.ONGOING:
        print("Game is ONGOING.")
